File: 1922_count_good_numbers.cpp
class Solution {
public:
    int countGoodNumbers(long long z) {
        return zZ(5, (z+1)/2) * 1LL * zZ(4, z/2) % O;
    }
private:
    const int O = 1e9+7;
    long long zZ(long long x, long long y) {
        long long o = 1;
        while(y) {
            if(y&1) o = o * x % O;
            x = x * x % O;
            y >>= 1;
        }
        return o;
    }
};

File: 1922_count_good_numbers.java
class Solution {
    public int countGoodNumbers(long a) {
        return (int)(q(5, (a+1)/2) * q(4, a/2) % M);
    }
    final long M = 1_000_000_007;
    long q(long b, long e) {
        long r = 1;
        while(e > 0) {
            if((e & 1) == 1) r = r * b % M;
            b = b * b % M;
            e >>= 1;
        }
        return r;
    }
}
